### DISCUSSION TITLE
CopilotChat fails to load despite manifest looking correct - please help!

### DISCUSSION THREAD

**Discussion by morrna** [CopilotC-Nvim/CopilotChat.nvim#141] [Q&A] [answered]
I installed the plugin manually. Here's the relevant piece of my `init.lua`:

```lua
local copilot_chat = require("CopilotChat")

copilot_chat.setup({
  debug = false,
  show_help = "yes",
  disable_extra_info = 'no',
  build = function()
    vim.notify("Please update the remote plugins by running ':UpdateRemotePlugins', then restart Neovim.")
  end,
  event = "VeryLazy",
})

-- Setup keymap
vim.keymap.set("n", "<leader>ccb", "<cmd>CopilotChatBuffer<cr>", { noremap = true })
vim.keymap.set("n", "<leader>cce", "<cmd>CopilotChatExplain<cr>", { noremap = true })
vim.keymap.set("n", "<leader>cct", "<cmd>CopilotChatTests<cr>", { noremap = true })
vim.keymap.set("x", "<leader>ccv", ":CopilotChatVisual<cr>", { noremap = true })
vim.keymap.set("x", "<leader>ccx", ":CopilotChatInPlace<cr>", { noremap = true })
```

Everything seems fine when I run `:UpdateRemotePlugins`:

```
remote/host: python3 host registered plugins ['CopilotChat']                                                                                                     
remote/host: generated rplugin manifest: /home/user/.local/share/nvim/rplugin.vim
```

I manually inspected the manifest file, and it looks just like the example.
When I restart neovim, I see this:

```
Error detected while processing /home/user/.config/nvim/init.lua:
E5113: Error while calling lua chunk: /home/user/.config/nvim/init.lua:24: module 'CopilotChat' not found:
        no field package.preload['CopilotChat']
        ...(more stack trace)
```

I've been spinning on this for a couple hours now, and I still have no leads. I've tried

- `:checkhealth`, which shows no problems with Python or pynvim.
- cd'ing into the rplugin/python3 directory, launching `python3 -i` and importing the CopilotChat module to make sure there are no import errors.
- inspecting the order of script loading with `nvim -V2`. Here I noticed that my configuration was being loaded before the manifest script, so I tried adding `vim.cmd("source /usr/share/nvim/runtime/autoload/remote/host.vim")` to the top of my `init.lua`. No change.

Something very strange seems to be going on. Any advice on what to try next would be appreciated!

Version info:

> NVIM v0.7.2
> Build type: Release
> LuaJIT 2.1.0-beta3
> Compiled by team+vim@tracker.debian.org
> 
> Features: +acl +iconv +tui
> See ":help feature-compile"

**Comment by jellydn** [comment]
Could you please try with a stable version? Your Neovim is a bit out of date.

**Comment by morrna** [comment]
I also tried everything above on neovim 0.10.0 from the [Neovim Unstable Debian PPA](https://launchpad.net/~neovim-ppa/+archive/ubuntu/unstable), and it works the same there. The version from my post is from the "stable" version. I could try other neovim versions for sure, but the fact that I saw the same error in both seems to point to a version-independent cause.

What version of neovim would you recommend? I'm also new to neovim (longtime vim user trying it out), so happy to take setup advice.

**Reply by jellydn** [reply]
> NVIM v0.7.2

> The version from my post is from the "stable" version

Could you try with v0.9.5 https://github.com/neovim/neovim/releases/tag/v0.9.5? Thanks.

**Reply by morrna** [reply]
Thanks!

So far I tried downloading it and running it alongside my other installation by calling it from the local path, and it gives the same error. Not sure if the fact that it's referencing the system scripts from the other installation would be an issue.

Based on the other thread, it looks like I was just missing a step that was omitted from the README. I'm going to give that a try first before changing my install around more.

**Comment by deathbeam** [comment]
Hmm this almost looks like path issue, e.g neovim cannot find the plugin:

```
E5113: Error while calling lua chunk: /home/user/.config/nvim/init.lua:24: module 'CopilotChat' not found:
        no field package.preload['CopilotChat']
```

are other plugins working for you?

**Reply by morrna** [reply]
The only other plugin I've installed so far is the regular Copilot plugin for completions. It's working fine.

I think I have the plugin module in the right place, so I would guess that it's on the path.
If you think of a way to debug this, I'm definitely curious.

Here's some truncated output showing where I put the plugin code:
```
:~/.config/nvim$ ls -R rplugin/

rplugin/:
python3

rplugin/python3:
CopilotChat

rplugin/python3/CopilotChat:
copilot_plugin.py
copilot.py
handlers
__init__.py
mypynvim
prompts.py
typings.py
utilities.py
```

**Reply by deathbeam** [reply]
oh did you only set up the rplugin part? As you also need the lua part of the plugin

**Reply by deathbeam** [reply]
Can also try canary branch which is pure lua so less confusing to manually setup maybe

**Reply by morrna** [reply]
The README.md of the main branch makes no mention of a lua part of the plugin, at least as far as I can see. Could you point me to it?

I could try the canary branch if I get 0.9.5 set up I suppose. How stable is it? Ideally I'd like to not have to mess with this much once I get up and running.

**Reply by deathbeam** [reply, marked as answer]
![image](https://github.com/CopilotC-Nvim/CopilotChat.nvim/assets/5115805/217146a7-f549-4480-bba8-0394df419633)

There is lua and rplugin directory, both required. The manual section in main branch definitely missing info for it tho but you can just copy the CopilotChat directory to ~/.config/nvim/lua/CopilotChat and it should work.

For canary, the API for it at least should be stable now and no more breaking changes are planned for it outside of bugfixes, but I am using it for a bit now and havent had any issues rly.

**Reply by deathbeam** [reply]
Also I would recommend putting plugins into ~/.config/nvim/pack/<name_of_pack>/start when not using package managers as that is the way you should do it instead of putting it to just ~/.config/nvim/lua and rplugin. E.g the recommended manual setup in main README.md is wrong. Which is fine most ppl just use package managers, but I also manage my installed plugins manually and the README.md on canary also has correct info now

**Reply by morrna** [reply]
Well, that graduated me to a new error at least!

```
Error detected while processing /home/user/.config/nvim/init.lua:
E5113: Error while calling lua chunk: Vim:E6100: "state" is not a valid stdpath
stack traceback:
        [C]: in function 'stdpath'
        /home/user/.config/nvim/lua/CopilotChat/vlog.lua:46: in function 'get_log_file'
        /home/user/.config/nvim/lua/CopilotChat/vlog.lua:52: in function 'new'
        /home/user/.config/nvim/lua/CopilotChat/vlog.lua:148: in main chunk
        [C]: in function 'require'
        /home/user/.config/nvim/lua/CopilotChat/utils.lua:3: in main chunk
        [C]: in function 'require'
        /home/user/.config/nvim/lua/CopilotChat/init.lua:1: in main chunk
        [C]: in function 'require'
        /home/user/.config/nvim/init.lua:23: in main chunk
Press ENTER or type command to continue
```

But trying with a locally downloaded nvim 0.9.5 seems to work! Guess I have to upgrade.

**Reply by morrna** [reply]
> Also I would recommend putting plugins into ~/.config/nvim/pack/<name_of_pack>/start when not using package managers as that is the way you should do it instead of putting it to just ~/.config/nvim/lua and rplugin. 

Could they live alongside the other copilot plugin within ~/.config/nvim/pack/github/start, or should I give them their own pack name?

**Reply by deathbeam** [reply]
By accident replied in separate message below but you should see it I hope

**Comment by deathbeam** [comment]
> > Also I would recommend putting plugins into ~/.config/nvim/pack/<name_of_pack>/start when not using package managers as that is the way you should do it instead of putting it to just ~/.config/nvim/lua and rplugin.
> 
> Could they live alongside the other copilot plugin within ~/.config/nvim/pack/github/start, or should I give them their own pack name?

Absolutely. The copilotchat in canary readme is just example, I personally use `~/.config/nvim/pack/bundle/start` for every plugin, you can just use `~/.config/nvim/pack/github/start` for everything too just fine as well. Example from my dotfiles: https://github.com/deathbeam/dotfiles/tree/master/nvim/.config/nvim/pack/bundle/start (I rly like to manage the plugins via submodules as well :d)

**Reply by morrna** [reply]
Thanks! That does look like a good idea.

**Reply by morrna** [reply]
I tried naively checking out the plugins into submodules like that without changing my init.lua, and for some reason UpdateRemotePlugins stopped finding the python piece of CopilotChat. Looks like I'll need to study up on some more details before I can make this sort of thing work. I definitely do appreciate you sharing your example, though!
